### ISSUE TITLE
Switch to `-Wswitch` instead of `-Wswitch-enum -Wswitch-default` for most code

### ISSUE THREAD

**Issue by encryptio** [rethinkdb/rethinkdb#5177]
We're currently using `-Wswitch-enum -Wswitch-default`, which requires that every switch on an enumeration explicitly mention each enumerated value _even if a default case exists_. Arguably, that's an over-restrictive default.

In several places, we do `switch (static_cast<int>(value)) {` to skip these restrictions, which IMO smells bad.

We do have certain `switch` statements where we do want _all_ enumerations _and_ a `default` case, and wish to raise errors if this is not the case. The `switch`es on `reql_version_t` and `cluster_version_t` are the main ones, because they need to handle unsanitized input (which is most cleanly done with a `default` case) and also _must_ be updated for every new enumeration we add (figuring out that an enumeration is missing after compilation succeeds is much more painful.)

@AtnNn, @VeXocide, and I discussed this on a draft code review (CR 3381), and settled on:

- Switching the compiler warning defaults to `-Wswitch` (removing `-Wswitch-enum -Wswitch-default`)
- Adding preprocessor macros to turn `-Wswitch-enum -Wswitch-default` on for those specific switch statements where they're helpful (@AtnNn made an example: https://gist.github.com/AtnNn/80e931a5d80551210830)
- Auditing all our `switch` statements to make sure that the macros are used when needed and no `static_cast<int>` hacks remain.

**Comment by mlucy** [comment]
So the new behavior would be that if a `default` branch isn't specified, the compiler still produces an error if you skip an `enum` value, but if a `default` branch is specified then it doesn't?

The only thing that makes me a tiny bit hesitant about this is that our current rule (where you put `default: unreachable()` at the end) fails quickly when you read from invalid memory rather than silently falling through.  (We end up reading enum values from freed or corrupted memory an embarassing amount, actually.)  But I suppose that's probably fine.

**Comment by encryptio** [comment]
Yes, having a `default` case will suppress an error about missing enumerations under `-Wswitch`.

**Comment by encryptio** [comment]
**tl;dr: I'm now of the opinion that this cure is worse than the disease, we should NOT do this, and stay with the slightly more painful but safer `-Wswitch-enum -Wswitch-default`.**

I worked on this a bit and partway through discussed some things with @mlucy;

The goals I had in mind for this were:
- Decrease the pain of partial `switch` statements
- Maintain compile-time errors on missing enumerations

But, while implementing this, I needed to remove a HUGE number of `default: unreachable()` cases because they'd conflict with the second goal under the new `-Wswitch` flag.

There are so many of these `default: unreachable()` cases that removing their runtime safety (to memory reuse, corruption, etc) to retain the same level of compile time safety under the new compiler flags seems far, far worse than doing the `switch (static_cast<int>(...))` hack in the cases where partial `switch`es are helpful.

In fact, @mlucy made a good point in person that a non-erroring `default` case (not `unreachable()`, `throw protocol_error`, etc) that doesn't raise compile time errors on missing enumerations is generally a scary thought in a codebase as large as ours is, since any person is unlikely to remember all the places an enumeration is `switch`ed on and if the `default` case is actually correct or not for the new value. Explicitly mentioning each case that has been thought about and having the compiler tell you about new ones is immensely valuable.

**Comment by danielmewes** [comment]
Ok that sounds reasonable.

Closing.
